sway/tree/container.c

```diff
@@ -1,61 +1,461 @@
+	struct sway_container *con = wl_container_of(
+			listener, con, output_enter);
+	struct wlr_scene_output *output = data;
+
+	if (con->view->foreign_toplevel) {
+		wlr_foreign_toplevel_handle_v1_output_enter(
+			con->view->foreign_toplevel, output->output);
+	}
+}
+
+static void handle_output_leave(
+		struct wl_listener *listener, void *data) {
+	struct sway_container *con = wl_container_of(
+			listener, con, output_leave);
+	struct wlr_scene_output *output = data;
+
+	if (con->view->foreign_toplevel) {
+		wlr_foreign_toplevel_handle_v1_output_leave(
+			con->view->foreign_toplevel, output->output);
+	}
+}
+
+static bool handle_point_accepts_input(
+		struct wlr_scene_buffer *buffer, double *x, double *y) {
+	return false;
+}
+
+static struct wlr_scene_rect *alloc_rect_node(struct wlr_scene_tree *parent,
+		bool *failed) {
+	if (*failed) {
+		return NULL;
+	}
+
+	// just pass in random values. These will be overwritten when
+	// they need to be used.
+	struct wlr_scene_rect *rect = wlr_scene_rect_create(
+		parent, 0, 0, (float[4]){0.f, 0.f, 0.f, 1.f});
+	if (!rect) {
+		sway_log(SWAY_ERROR, "Failed to allocate a wlr_scene_rect");
+		*failed = true;
+	}
+
+	return rect;
+}
+
 struct sway_container *container_create(struct sway_view *view) {
 	struct sway_container *c = calloc(1, sizeof(struct sway_container));
 	if (!c) {
 		sway_log(SWAY_ERROR, "Unable to allocate sway_container");
 		return NULL;
 	}
 	node_init(&c->node, N_CONTAINER, c);
-	c->pending.layout = L_NONE;
-	c->view = view;
-	c->alpha = 1.0f;
-	c->saturation = 1.0f;
-	c->dim = config->default_dim_inactive;
-	c->shadow_enabled = config->shadow_enabled;
-	c->blur_enabled = config->blur_enabled;
-	c->corner_radius = config->corner_radius;
+
+	// TODO: add shadow to below
+	// Container tree structure
+	// - scene tree
+	//   - title bar
+	//     - border
+	//     - background
+	//     - title text
+	//     - marks text
+	//   - border
+	//     - border top/bottom/left/right
+	//     - content_tree (we put the content node here so when we disable the
+	//       border everything gets disabled. We only render the content iff there
+	//       is a border as well)
+	//     - buffer used for output enter/leave events for foreign_toplevel
+	bool failed = false;
+	c->scene_tree = alloc_scene_tree(root->staging, &failed);
+
+	c->title_bar.tree = alloc_scene_tree(c->scene_tree, &failed);
+
+	c->border.tree = alloc_scene_tree(c->scene_tree, &failed);
+	c->content_tree = alloc_scene_tree(c->border.tree, &failed);
+
+	if (view) {
+		// only containers with views can have borders
+		c->border.top = alloc_rect_node(c->border.tree, &failed);
+		c->border.bottom = alloc_rect_node(c->border.tree, &failed);
+		c->border.left = alloc_rect_node(c->border.tree, &failed);
+		c->border.right = alloc_rect_node(c->border.tree, &failed);
+
+		c->title_bar.border = alloc_rect_node(c->title_bar.tree, &failed);
+		c->title_bar.background = alloc_rect_node(c->title_bar.tree, &failed);
```

Comment:
Initializing here currently causes a segfault when moving tiled views. It's easiest to always allocate all nodes and just hide them if they're not being used :)